src/connectors/impls/gcs/api_client.rs

```diff
@@ -0,0 +1,476 @@
+            HttpTaskCommand::UploadData { file, data } => {
+                self.upload_data(done_until, file, data).await
+            }
+        }
+    }
+}
+
+impl<THttpClient: HttpClient> DefaultApiClient<THttpClient> {
+    #[cfg(test)]
+    pub fn new(client: THttpClient) -> Result<Self> {
+        Ok(Self {
+            sessions_per_file: HashMap::new(),
+            client,
+        })
+    }
+
+    #[cfg(not(test))]
+    pub fn new(client: THttpClient) -> Result<Self> {
+        Ok(Self {
+            token: Token::new()?,
+            sessions_per_file: HashMap::new(),
+            client,
+        })
+    }
+
+    async fn upload_data(
+        &mut self,
+        done_until: Arc<AtomicUsize>,
+        file: FileId,
+        data: BufferPart,
+    ) -> Result<()> {
+        let mut response = None;
+        for i in 0..3 {
+            let session_url =
+                self.sessions_per_file
+                    .get(&file)
+                    .ok_or(ErrorKind::GoogleCloudStorageError(
+                        "No session URL is available",
+                    ))?;
+            let mut request = Request::new(Method::Put, session_url.clone());
+
+            request.insert_header(
+                "Content-Range",
+                format!(
+                    "bytes {}-{}/*",
+                    data.start,
+                    // -1 on the end is here, because Content-Range is inclusive and our range is exclusive
+                    data.start + data.len() - 1
+                ),
+            );
+            request.insert_header("User-Agent", "Tremor");
+            request.insert_header("Accept", "*/*");
+            request.set_body(data.data.clone());
+
+            match self.client.send(request).await {
+                Ok(request) => response = Some(request),
+                Err(e) => {
+                    warn!("Failed to send a request to GCS: {}", e);
+
+                    sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+                    continue;
+                }
+            }
+
+            if let Some(response) = response.as_mut() {
+                if !response.status().is_server_error() && response.header("range").is_some() {
+                    break;
+                }
+
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+
+                sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+            }
+        }
+
+        if let Some(mut response) = response {
+            if response.status().is_server_error() {
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+                return Err("Received server errors from Google Cloud Storage".into());
+            }
+
+            if let Some(raw_range) = response.header("range") {
```

Comment:
This is nicer written as:

```rust
let new_range = response.header("range").ok_or("No range header")?;
```

it removes a level of nesting and puts the error right next to where it occurs instead of 20 lines below :)